### ISSUE TITLE
Proper indexed route reverse

### ISSUE THREAD

**Issue by jtraub** [starlite-api/starlite#540]
**What's the feature you'd like to ask for.**
Sometimes I need to redirect client app to another endpoint. To do that I need to generate an url for this endpoint.

Starlite provides `get_handler_index_by_name` which returns a dict with original path but it can not be used as an url path in case it contains path parameters. So in order to create an url to a handler like this

```python
@get(path="/complex/{param:int}/path/{param2:str}/", name="complex-path")
def handler(param: int, param2: str) -> str:
    ...
```

user would have to parse path and replace parameters with actual values.

I think it would be very nice to have a function like Django's [`reverse`](https://docs.djangoproject.com/en/4.1/ref/urlresolvers/#reverse) which would accept route index name and path parameters as kwargs.

**Comment by jtraub** [comment]
My attempt to implement such function can be seen here

https://github.com/jtraub/starlite/compare/main...jtraub:starlite:route_reverse

I think included tests give an idea about function behaviour

**Comment by Goldziher** [comment]
That's pretty cool. Wanna add a PR?

Also, check this bit if code out:
https://github.com/starlite-api/starlite/blob/main/starlite/router.py#L198

This was added by @infohash , and it's a full mapping of all route handlers, via their `handler.fn.__qualname__` to paths. It's a better source than the name index, because it doesn't rely on names as such. But it's not exposed via a method that retrieves the path, it's just a mapping for now. 

We need a solution that normalizes these two tbh.

**Comment by jtraub** [comment]
> That's pretty cool. Wanna add a PR?

Sure. I will add docs tomorrow and then submit a PR.

>  full mapping of all route handlers, via their handler.fn.__qualname__ to paths.

I saw it when I went through sources.
I still think that names are better than FQN in such cases because names can be short, concise (compare `group_info` vs `app.core.api.group.info`) and do not depend on codebase organization. Another benefit of named resources is that this concept is familiar to developers coming from other frameworks (aiohttp, fastapi, django, flask and many others). 

However, merging this piece of code and route names together would allow to resolve a few problems.
* named routes can not be registered multiple times under different routers
* named routes do not work with optional path parameters

**Comment by Goldziher** [comment]
Ok, go ahead and merge these. We dont need this on the router level IMO. Please correct me if im wrong @infohash - but we can have the logic only on the app instance, and have a single method for it.

**Comment by infohash** [comment]
@Goldziher That's correct. Starlite instance can see all the registered route handlers and the user can also access app instance inside a route handler by using `Request.app` so the public method must be on the app instance.

**Comment by jtraub** [comment]
Ah, damn, I missed the last comments.

I can do the merge but I would like to take some time here and do experiments as I want to be sure most corner cases with route indexing has been covered.

At the moment I am only struggling with optional path parameters on my pet project though.

**Comment by Goldziher** [comment]
> Ah, damn, I missed the last comments.
> 
> I can do the merge but I would like to take some time here and do experiments as I want to be sure most corner cases with route indexing has been covered.
> 
> At the moment I am only struggling with optional path parameters on my pet project though.

Take your time
